code/game/machinery/kitchen/microwave.dm

```diff
@@ -324,18 +283,60 @@
 		for (var/obj/item/weapon/reagent_containers/food/snacks/S in contents)
 			S.cook()
 
-		dispose(0) //clear out anything left
-		stop()
+		eject(0) //clear out anything left
 
 		return
 
-/obj/machinery/microwave/proc/wzhzhzh(var/seconds as num) // Whoever named this proc is fucking literally Satan. ~ Z
-	for (var/i=1 to seconds)
-		if (stat & (NOPOWER|BROKEN))
-			return 0
-		use_power(active_power_usage)
-		sleep(10)
-	return 1
+/obj/machinery/microwave/proc/cook_for_time(var/cook_time = 20, var/failed = FALSE, var/dirt = FALSE, var/broke = FALSE) // Formerly proc/wzhzhzh, may it rest in peace
+	start(dirt)
+
+	var/half_time = cook_time / 2
+	var/obj/cooked
+	for (var/i=1 to half_time)
+		if ((stat & (NOPOWER|BROKEN)) || abort) // Safe to abort early on
+			stop(dirt, broke)
+			return FALSE
+		else
+			use_power(active_power_usage)
+			update_cooking_ui(cook_time, i)
+			sleep(10)
+
+	playsound(src, 'sound/machines/click.ogg', 20, 1) // Mid-way through cooking 'click'
+
+	for (var/j=half_time to cook_time)
+		if ((stat & (NOPOWER|BROKEN)))
+			stop()
+			return FALSE
+		else if(abort) // Not safe to abort anymore - egads, the roast is ruined!
+			if(failed)
+				cooked = fail()
+				cooked.forceMove(loc)
+			stop(dirt, broke)
+			return FALSE
+		else
+			use_power(active_power_usage)
+			update_cooking_ui(cook_time, j)
+			sleep(10)
+
+	if(failed)
+		cooked = fail()
+		cooked.forceMove(loc)
+		stop(dirt, broke)
+		return FALSE
+
+	stop(dirt, broke)
+
+	return TRUE
+
+/obj/machinery/microwave/proc/update_cooking_ui(var/cook_time, var/cur_time)
+	for(var/datum/vueui/ui in SSvueui.get_open_uis(src))
+		ui.data["cook_time"] = cook_time
+		ui.data["cur_time"] = cur_time
+	SSvueui.check_uis_for_change(src)

```

Comment:
You should not directly modify in this case ui data. Use var monitors.

In this case you should just keep track of when cooking started and when it should end. Use `world.time` DM and `wtime` state variable for time tracking. For how it can be used, look at faxmachine code.

Reply:
Didn't even think of that, and you're totally right. Hopefully the new version is a bit better.